src/text-input/index.ts

```diff
@@ -166,13 +202,14 @@ export class TextInputBase<P extends TextInputProperties = TextInputProperties>
 			readOnly,
 			required
 		} = this.properties;
+		const { valid } = this._state;
 		const focus = this.meta(Focus).get('root');
 		return [
 			css.root,
 			disabled ? css.disabled : null,
 			focus.containsFocus ? css.focused : null,
-			invalid === true ? css.invalid : null,
-			invalid === false ? css.valid : null,
+			invalid === true || valid === false ? css.invalid : null,
```

Comment:
If we're adding this we will want to remove support for passing `invalid` manually.